src/biz/tugay/groovyship/service/BoardService.groovy

```diff
@@ -0,0 +1,56 @@
+package biz.tugay.groovyship.service
+
+import biz.tugay.groovyship.modal.Board
+import biz.tugay.groovyship.modal.Ship
+import biz.tugay.groovyship.modal.Coordinate
+
+class BoardService
+{
+
+  ShipService shipService = new ShipService()
+
+  /**
+   * Attempts to add the {@code ship} to the board. As per the rules of the game, a ship cannot overlay with
+   * any other existing ship on the board. Ships also cannot be adjacent to each other.
+   *
+   * @param board The board the ship will be added to, potentially having ships already added.
+   * @param ship The new ship that is being attempted to be added.
+   * @return Whether ship was added to the board or not.
+   */
+  boolean addShip(Board board, Ship ship) {
+    def shipCoordinates = ship.coordinateIsHitByMissileMap.keySet()
+
+    if (board.ships.any { shipService.isBoundariesOfShipInCoordinates it, shipCoordinates }) {
```

Comment:
`any`, `every`, `each`, `find` and so on are powerful methods that accept a method. Similar to streams in Java. `it` is implicitly the only argument passed in.

Reply:
i think similar again to python:
there is no dict.exists(key) but `if key in dict.keys() `

Reply:
`if key in dict.keys()` sounds different to me. `any` is accepting a method which it will invoke for each of the element of the list it was called on, and returns `true` if any of the method invocations return `true`.